services/src/datasets/external/netcdfcf/mod.rs

```diff
@@ -0,0 +1,1109 @@
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+use crate::datasets::listing::DatasetListOptions;
+use crate::datasets::listing::{ExternalDatasetProvider, ProvenanceOutput};
+use crate::error::Error;
+use crate::{
+    datasets::{listing::DatasetListing, storage::ExternalDatasetProviderDefinition},
+    util::user_input::Validated,
+};
+use async_trait::async_trait;
+use chrono::NaiveDate;
+use gdal::cpl::CslStringList;
+use gdal::errors::GdalError;
+use gdal::{Dataset, DatasetOptions, GdalOpenFlags};
+use gdal_sys::{
+    CPLErr, GDALAttributeReadAsString, GDALDataType, GDALDatasetGetRootGroup, GDALDimensionGetSize,
+    GDALExtendedDataTypeGetNumericDataType, GDALGroupGetAttribute, GDALGroupGetGroupNames,
+    GDALGroupHS, GDALGroupOpenGroup, GDALGroupOpenMDArray, GDALMDArrayGetAttribute,
+    GDALMDArrayGetDataType, GDALMDArrayGetDimensions, GDALMDArrayGetNoDataValueAsDouble,
+    GDALMDArrayGetSpatialRef, GDALMDArrayH,
+};
+use geoengine_datatypes::dataset::{DatasetId, DatasetProviderId, ExternalDatasetId};
+use geoengine_datatypes::primitives::{
+    Measurement, RasterQueryRectangle, TimeGranularity, TimeInstance, TimeStep,
+    VectorQueryRectangle,
+};
+use geoengine_datatypes::raster::{GdalGeoTransform, RasterDataType};
+use geoengine_datatypes::spatial_reference::{SpatialReference, SpatialReferenceOption};
+use geoengine_operators::engine::TypedResultDescriptor;
+use geoengine_operators::source::{
+    FileNotFoundHandling, GdalDatasetGeoTransform, GdalDatasetParameters, GdalMetadataNetCdfCf,
+};
+use geoengine_operators::util::gdal::gdal_open_dataset_ex;
+use geoengine_operators::{
+    engine::{MetaData, MetaDataProvider, RasterResultDescriptor, VectorResultDescriptor},
+    mock::MockDatasetDataSourceLoadingInfo,
+    source::{GdalLoadingInfo, OgrSourceDataset},
+};
+use log::debug;
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+
+pub use self::error::NetCdfCf4DProviderError;
+
+mod error;
+
+type Result<T, E = NetCdfCf4DProviderError> = std::result::Result<T, E>;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct NetCdfCfDataProviderDefinition {
+    pub id: DatasetProviderId,
+    pub name: String,
+    pub path: PathBuf,
+}
+
+#[typetag::serde]
+#[async_trait]
+impl ExternalDatasetProviderDefinition for NetCdfCfDataProviderDefinition {
+    async fn initialize(self: Box<Self>) -> crate::error::Result<Box<dyn ExternalDatasetProvider>> {
+        Ok(Box::new(NetCdfCfDataProvider {
+            id: self.id,
+            path: self.path,
+        }))
+    }
+
+    fn type_name(&self) -> String {
+        "NetCdfCfProviderDefinition".to_owned()
+    }
+
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn id(&self) -> DatasetProviderId {
+        self.id
+    }
+}
+
+pub struct NetCdfCfDataProvider {
+    id: DatasetProviderId,
+    path: PathBuf,
+}
+
+/// TODO: This should be part of the GDAL crate
+/// TODO: free group with `GDALGroupRelease()`
```

Comment:
i guess this is a leak then?

Reply:
Yes

Reply:
I removed all leaks and checked the tests with ASAN:

`RUSTFLAGS="-Z sanitizer=leak" cargo test --package geoengine-services --lib --all-features --target x86_64-unknown-linux-gnu -- datasets::external::netcdfcf::tests`

e109efe